// @netlify/db-dev/dist/main.js
// src/main.ts
import { mkdir } from "fs/promises";
import { createServer as createNetServer } from "net";
import { PGlite } from "@electric-sql/pglite";
import { fromNodeSocket } from "pg-gateway/node";
var DEFAULT_HOST = "localhost";
var NetlifyDB = class {
  db;
  directory;
  logger;
  port;
  server;
  constructor({ directory, logger, port } = {}) {
    this.directory = directory;
    this.logger = logger ?? console.log;
    this.port = port;
  }
  async start() {
    if (this.directory) {
      await mkdir(this.directory, { recursive: true });
    }
    this.db = await PGlite.create(this.directory);
    this.server = createNetServer((socket) => {
      void this.handleConnection(socket);
    });
    return new Promise((resolve, reject) => {
      if (!this.server) {
        reject(new Error("Server not initialized"));
        return;
      }
      this.server.on("error", reject);
      this.server.listen(this.port ?? 0, DEFAULT_HOST, () => {
        this.server?.off("error", reject);
        const { address, port } = this.server?.address();
        const host = address === "::1" || address === "127.0.0.1" ? "localhost" : address;
        resolve(`postgres://${host}:${String(port)}/postgres`);
      });
    });
  }
  async stop() {
    return new Promise((resolve, reject) => {
      if (!this.server) {
        resolve();
        return;
      }
      this.server.close((err) => {
        if (err) {
          reject(err);
        } else {
          resolve();
        }
      });
    });
  }
  async handleConnection(socket) {
    if (!this.db) {
      return;
    }
    const db = this.db;
    try {
      await fromNodeSocket(socket, {
        serverVersion: "16.3 (NetlifyDB/pglite)",
        auth: {
          method: "trust"
        },
        async onMessage(data, { isAuthenticated }) {
          if (!isAuthenticated) {
            return;
          }
          return db.execProtocolRaw(data);
        }
      });
    } catch (error) {
      if (error instanceof Error && error.message.includes("ECONNRESET")) {
        return;
      }
      this.logger("Unexpected connection error:", error);
    }
  }
};
export {
  NetlifyDB
};
